File: src/shm.zig
const std = @import("std");
const c = @cImport({
    @cInclude("sys/mman.h");
    @cInclude("sys/stat.h");
    @cInclude("fcntl.h");
    @cInclude("unistd.h");
    @cInclude("errno.h");
});

// Export C-compatible functions for Bun FFI

export fn ipc_shm_open(name_ptr: [*]const u8, name_len: usize, size: usize) i32 {
    const name = name_ptr[0..name_len];
    // We need a null-terminated string for C API.
    // In a real app we might allocate, but here we expect the caller to provide a safe name.
    // However, shm_open expects a C string.
    // Let's alloc a buffer on stack or heap.
    var buf: [256]u8 = undefined;
    if (name.len >= buf.len - 1) return -1;
    @memcpy(buf[0..name.len], name);
    buf[name.len] = 0;
    const c_name = @as([*c]const u8, &buf);

    // O_CREAT | O_RDWR
    const fd = c.shm_open(c_name, c.O_CREAT | c.O_RDWR, 0o666);
    if (fd < 0) return -1;

    // Truncate to size
    const res = c.ftruncate(fd, @intCast(size));
    if (res < 0) {
        _ = c.close(fd);
        _ = c.shm_unlink(c_name);
        return -1;
    }

    return fd;
}

export fn ipc_shm_unlink(name_ptr: [*]const u8, name_len: usize) i32 {
    var buf: [256]u8 = undefined;
    if (name_len >= buf.len - 1) return -1;
    @memcpy(buf[0..name_len], name_ptr[0..name_len]);
    buf[name_len] = 0;
    const c_name = @as([*c]const u8, &buf);

    return c.shm_unlink(c_name);
}

// Map the memory
// Returns pointer to memory or null on failure
export fn ipc_mmap(fd: i32, size: usize) ?*anyopaque {
    const ptr = c.mmap(null, size, c.PROT_READ | c.PROT_WRITE, c.MAP_SHARED, fd, 0);
    if (ptr == c.MAP_FAILED) return null;
    return ptr;
}

export fn ipc_munmap(ptr: *anyopaque, size: usize) i32 {
    return c.munmap(ptr, size);
}

export fn ipc_close(fd: i32) i32 {
    return c.close(fd);
}

File: src/shm.c
#include <sys/mman.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>

// Export C-compatible functions for Bun FFI

int ipc_shm_open(const char* name, size_t name_len, size_t size) {
    // We need a null-terminated string.
    // The caller (Bun) sends a pointer and length.
    // We can assume it's null-terminated if Bun sends CString?
    // But Bun FFI 'ptr' with buffer might not be null terminated if not careful.
    // We'll copy to be safe.
    char name_buf[256];
    if (name_len >= 255) return -1;
    memcpy(name_buf, name, name_len);
    name_buf[name_len] = '\0';

    int fd = shm_open(name_buf, O_CREAT | O_RDWR, 0666);
    if (fd < 0) return -1;

    if (ftruncate(fd, (off_t)size) < 0) {
        close(fd);
        shm_unlink(name_buf);
        return -1;
    }

    return fd;
}

int ipc_shm_unlink(const char* name, size_t name_len) {
    char name_buf[256];
    if (name_len >= 255) return -1;
    memcpy(name_buf, name, name_len);
    name_buf[name_len] = '\0';
    
    return shm_unlink(name_buf);
}

void* ipc_mmap(int fd, size_t size) {
    void* ptr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    if (ptr == MAP_FAILED) return NULL;
    return ptr;
}

int ipc_munmap(void* ptr, size_t size) {
    return munmap(ptr, size);
}

int ipc_close(int fd) {
    return close(fd);
}
